BParkingNano/plugins/DiLeptonBuilder.cc

```diff
@@ -0,0 +1,124 @@
+#include "CommonTools/Statistics/interface/ChiSquaredProbability.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrack.h"
+#include "helper.h"
+#include <limits>
+#include <algorithm>
+#include "KinVtxFitter.h"
+
+template<typename Lepton>
+class DiLeptonBuilder : public edm::global::EDProducer<> {
+
+  // perhaps we need better structure here (begin run etc)
+public:
+  typedef std::vector<Lepton> LeptonCollection;
+  typedef std::vector<reco::TransientTrack> TransientTrackCollection;
+
+  explicit DiLeptonBuilder(const edm::ParameterSet &cfg):
+    l1_selection_{cfg.getParameter<std::string>("lep1Selection")},
+    l2_selection_{cfg.getParameter<std::string>("lep2Selection")},
+    pre_vtx_selection_{cfg.getParameter<std::string>("preVtxSelection")},
+    post_vtx_selection_{cfg.getParameter<std::string>("postVtxSelection")},
+    src_{consumes<LeptonCollection>( cfg.getParameter<edm::InputTag>("src") )},
+    ttracks_src_{consumes<TransientTrackCollection>( cfg.getParameter<edm::InputTag>("transientTracksSrc") )} {
+       produces<pat::CompositeCandidateCollection>();
+    }
+
+  ~DiLeptonBuilder() override {}
+  
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions) {}
+  
+private:
+  const StringCutObjectSelector<Lepton> l1_selection_; // cut on leading lepton
+  const StringCutObjectSelector<Lepton> l2_selection_; // cut on sub-leading lepton
+  const StringCutObjectSelector<pat::CompositeCandidate> pre_vtx_selection_; // cut on the di-lepton before the SV fit
+  const StringCutObjectSelector<pat::CompositeCandidate> post_vtx_selection_; // cut on the di-lepton after the SV fit
+  const edm::EDGetTokenT<LeptonCollection> src_;
+  const edm::EDGetTokenT<TransientTrackCollection> ttracks_src_;
+};
+
+template<typename Lepton>
+void DiLeptonBuilder<Lepton>::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const &) const {
+
+  //input
+  edm::Handle<LeptonCollection> leptons;
+  evt.getByToken(src_, leptons);
+  
+  edm::Handle<TransientTrackCollection> ttracks;
+  evt.getByToken(ttracks_src_, ttracks);
+
+  // output
+  std::unique_ptr<pat::CompositeCandidateCollection> ret_value(new pat::CompositeCandidateCollection());
+  
+  for(size_t l1_idx = 0; l1_idx < leptons->size(); ++l1_idx) {
+    edm::Ptr<Lepton> l1_ptr(leptons, l1_idx);
+    bool l1_as_l1 = l1_selection_(*l1_ptr);
+    bool l1_as_l2 = l2_selection_(*l1_ptr);
+    if( !(l1_as_l1 || l1_as_l2) ) continue; 
+    
+    for(size_t l2_idx = l1_idx + 1; l2_idx < leptons->size(); ++l2_idx) {
+      edm::Ptr<Lepton> l2_ptr(leptons, l2_idx);
+      bool l2_as_l1 = l1_selection_(*l2_ptr);
+      bool l2_as_l2 = l2_selection_(*l2_ptr);
+      if( !(l2_as_l1 || l2_as_l2) ) continue;
+
+      bool l1_l2 = l1_as_l1 && l2_as_l2;
+      bool l2_l1 = l2_as_l1 && l1_as_l2;
+      if( !(l1_l2 || l2_l1) ) continue; // make sure both leptons pass opposite selections
```

Comment:
All this from L70 can be simplified for pt sorted collections
- already the case for pat::Muons, pat::Muons
- soon the case for pat::Ele, pat::Ele
- valid also for pat::Muon, track    with l1 from pat::Muon only,  and simply checking that pt2 < pt1 when considering the track as l2